File: python/15_bsearch/bsearch.py
"""
    Author: Wenru
"""

from typing import List

def bsearch(nums: List[int], target: int) -> int:
    """Binary search of a target in a sorted array
    without duplicates. If such a target does not exist,
    return -1, othewise, return its index.
    """
    low, high = 0, len(nums) - 1
    while low <= high:
        mid = low + (high - low) // 2
        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            low = mid + 1
        else:
            high = mid - 1
    
    return -1

File: scala/src/main/scala/15_bsearch/BSearch.scala
object BSearch {
    def search(nums: Array[Int], target: Int): Int = {
        var low = 0
        var high = nums.length - 1
        while(low <= high){
            val mid = low + ((high - low) >> 2)
            if(nums(mid) > target){
                high = mid - 1
            } else if (nums(mid) < target){
                low = mid + 1
            } else {
                return mid
            }
        }
        
        return -1
    }
}
